src/linux/ptrace_dumper.rs

```diff
@@ -101,23 +187,43 @@ impl PtraceDumper {
             mappings: Vec::new(),
             page_size: 0,
         };
-        dumper.init(stop_timeout)?;
+        dumper.init(stop_timeout, soft_errors)?;
         Ok(dumper)
     }
 
     // TODO: late_init for chromeos and android
-    pub fn init(&mut self, stop_timeout: Duration) -> Result<(), InitError> {
+    pub fn init(
+        &mut self,
+        stop_timeout: Duration,
+        mut soft_errors: impl WriteErrorList<InitError>,
+    ) -> Result<(), InitError> {
         // Stopping the process is best-effort.
         if let Err(e) = self.stop_process(stop_timeout) {
-            log::warn!("failed to stop process {}: {e}", self.pid);
+            soft_errors.push(InitError::StopProcessFailed(e));
         }
 
-        if let Err(e) = self.auxv.try_filling_missing_info(self.pid) {
-            log::warn!("failed trying to fill in missing auxv info: {e}");
+        // Even if we completely fail to fill in any additional Auxv info, we can still press
+        // forward.
+        if let Err(e) = self.auxv.try_filling_missing_info(
+            self.pid,
+            soft_errors.subwriter(InitError::FillMissingAuxvInfoErrors),
+        ) {
+            soft_errors.push(InitError::FillMissingAuxvInfoFailed(e));
+        }
```

Comment:
Just an idea here. Maybe it would be worth adding a function to  the list of errors so that this could be written as:
```rust
let res = self.auxv.try_filling_missing_info(self.pid, soft_errors.subwriter(InitError::FillMissingAuxvInfoErrors));
soft_errors.record_if_err(res); // or maybe_record_err(), whatever feels right
```

Reply:
Oooo! I always like suggestions to improve ergonomics!

Your example is a good start,  but I think it only works as-written in this specific case, where the return type of the function is `Result<(), E>`. If we want it to be more generic to functions that return `Result<T, E>`, the API will need to be a bit more complex to pass forward the `Ok(T)` value.

For example, in our API above if `try_filling_missing_info()` actually returned a value (let's say `Result<Foo, AuxvError>`), we'd want our API to: map `Ok(Foo)` to `Foo`, map `Err(AuxvError)` to `AuxvError`, then map it with a user function to `InitError`, then push that onto the list, and finally return a `Foo` type.

So, I guess to do that, we'd have to have our new API take 2 functions: A function that maps `AuxvError` to `InitError` (which we already have, `InitError::FillMissingAuxvInfoFailed`) and a function that takes no arguments and produces a `Foo` (similar to `Option::unwrap_or_else()`). Normally the latter would just be `Foo::default` (We could also have a function similar to `unwrap_or_default()` if we want).

So, our new function (added to `trait WriteErrorList<E>`) would look something like this:

```Rust
fn unwrap_or_record_and_then<T, SubErr, SubMapFn, AndThenFn>(
    &mut self,
    result: Result<T, SubErr>,
    map_fn: SubMapFn,
    and_then: AndThenFn,
) -> T
where
    SubMapFn: FnOnce(SubErr) -> E,
    AndThenFn: FnOnce() -> T,
{
    match result {
        Ok(t) => t,
        Err(e) => {
            let parent_error = map_fn(e);
            self.push(parent_error);
            and_then()
        }
    }
}
```

And then here's what it would look like when it's used:

```Rust
let res = self.auxv.try_filling_missing_info(
    self.pid,
    soft_errors.subwriter(InitError::FillMissingAuxvInfoErrors),
);
let foo = soft_errors.unwrap_or_record_and_then(
    res,
    InitError::FillMissingAuxvInfoFailed,
    Foo::default);
```

Which just becomes your code example above in the special case that `T = ()`.

What do you think? Does it seem worth adding?

Reply:
Hmm... There is another way that might be even-more ergonomic -- Implement a trait directly on the `Result<T, E>` that allows recording it into a soft error. Then the code would look like this:

```Rust
let foo = self.auxv.try_filling_missing_info(
    self.pid,
    soft_errors.subwriter(InitError::FillMissingAuxvInfoErrors),
)
.unwrap_or_record_and_default(&mut soft_errors, InitError::FillMissingAuxvInfoFailed);
```

What do you think? Should we do the first one, this one, or neither?

Reply:
`unwrap_or_record_and_default()` makes this very readable IMHO. Let's keep it for a subsequent iteration though, since you'd need to update the other crates too.